File: LLM-Translations/GPT-4/codenet/C/Java/s808208135.java
import java.util.Scanner;

class Date {
    int go, back, cost;
}

public class s808208135 {
    static int N = 100000;
    static int INF = 1000000;
    static int n;
    static int[] d = new int[N];
    static Date[] a = new Date[500000];

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        n = in.nextInt();
        int count = 0;

        for (int i = 0; i < n; i++) {
            int u = in.nextInt();
            int k = in.nextInt();
            for (int j = 0; j < k; j++) {
                a[count] = new Date();
                a[count].back = u;
                a[count].go = in.nextInt();
                a[count].cost = in.nextInt();
                count++;
            }
        }
        root(count);
        for (int i = 0; i < n; i++)
            System.out.println(i + " " + d[i]);
    }

    public static void root(int b) {
        int flag = 1;

        for (int i = 0; i < n; i++)
            d[i] = INF;
        d[0] = 0;

        while (flag != 0) {
            flag = 0;
            for (int i = 0; i < b; i++) {
                if (a[i].cost + d[a[i].back] < d[a[i].go]) {
                    d[a[i].go] = a[i].cost + d[a[i].back];
                    flag = 1;
                }
            }
        }
    }
}
// 

File: LLM-Translations/GPT-4/codenet/C/Python/s808208135.py

from sys import stdin

N = 100000
INF = 1000000

class Date:
    def __init__(self, go, back, cost):
        self.go = go
        self.back = back
        self.cost = cost

def root(a, b):
    d = [INF] * n
    d[0] = 0
    flag = 1

    while flag != 0:
        flag = 0
        for i in range(b):
            if a[i].cost + d[a[i].back] < d[a[i].go]:
                d[a[i].go] = a[i].cost + d[a[i].back]
                flag = 1
                
    return d

def main():
    n = int(input().strip())
    a = []
    count = 0
    for _ in range(n):
        u, k, *edges = map(int, input().strip().split())
        for j in range(0, len(edges), 2):
            go, cost = edges[j], edges[j+1]
            a.append(Date(go, u, cost))
            count += 1
            
    d = root(a, count)
    for i in range(n):
        print(i, d[i])

if __name__ == "__main__":
    main()


#
